interceptor/zipkin/trace_interceptor.go

```diff
@@ -0,0 +1,399 @@
+// Copyright 2018, OpenCensus Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package zipkininterceptor
+
+import (
+	"compress/gzip"
+	"compress/zlib"
+	"context"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"strings"
+
+	"go.opencensus.io/trace"
+
+	zipkinmodel "github.com/openzipkin/zipkin-go/model"
+
+	commonpb "github.com/census-instrumentation/opencensus-proto/gen-go/agent/common/v1"
+	agenttracepb "github.com/census-instrumentation/opencensus-proto/gen-go/agent/trace/v1"
+	tracepb "github.com/census-instrumentation/opencensus-proto/gen-go/trace/v1"
+	"github.com/census-instrumentation/opencensus-service/interceptor"
+	"github.com/census-instrumentation/opencensus-service/internal"
+	"github.com/census-instrumentation/opencensus-service/spanreceiver"
+)
+
+type ZipkinInterceptor struct {
+	spanSink spanreceiver.SpanReceiver
+}
+
+var _ interceptor.TraceInterceptor = (*ZipkinInterceptor)(nil)
+var _ http.Handler = (*ZipkinInterceptor)(nil)
+
+func New(sr spanreceiver.SpanReceiver) (*ZipkinInterceptor, error) {
+	return &ZipkinInterceptor{spanSink: sr}, nil
+}
+
+func (zi *ZipkinInterceptor) StartTraceInterception(ctx context.Context, spanSink spanreceiver.SpanReceiver) error {
+	zi.spanSink = spanSink
+	return nil
+}
+
+func (zi *ZipkinInterceptor) parseAndConvertToTraceSpans(jsonBlob []byte) (reqs []*agenttracepb.ExportTraceServiceRequest, err error) {
+	var zipkinSpans []*zipkinmodel.SpanModel
+	if err = json.Unmarshal(jsonBlob, &zipkinSpans); err != nil {
+		return nil, err
+	}
+
+	// *commonpb.Node instances have unique addresses hence
+	// for grouping within a map, we'll use the .String() value
+	byNodeGrouping := make(map[string][]*tracepb.Span)
+	nodes := make([]*commonpb.Node, 0, len(zipkinSpans))
+	// Now translate them into tracepb.Span
+	for _, zspan := range zipkinSpans {
+		span, node, err := zipkinSpanToTraceSpan(zspan)
+		// TODO:(@odeke-em) record errors
+		if err == nil && span != nil {
+			key := node.String()
+			if _, firstTime := byNodeGrouping[key]; firstTime {
```

Comment:
This is appending nodes with same key after the first time that the key is processed. So if the key appears only once it won't be present on nodes, which gives the impression that it should have been: 
```go
if _, visited := byNodeGrouping[key]; !visited {
```

Reply:
Great catch, thanks @pjanotti!